File: HID/GHP.cpp
#include "GHP.h"

const byte DEVICE_ID = 53;

const byte ANALOG_CHANGE = 'A';
const byte BUTTON_PRESS = 'B';
const byte BUTTON_RELEASE = 'R';
const byte ENCODER_INCREMENT = 'I';
const byte ENCODER_DECREMENT = 'D';
const byte RESPONCE_PERSISTENT = 'r';
const byte RESPONCE_DISOVER = 'w';

const byte SYSTEM = 'S';
const byte SYSTEM_DISCOVER = 'W';
const byte SYSTEM_RESET = 'R';

const byte SET_LED = 'L';
const byte SET_LCD = 'l';
const byte SET_PERISTENT = 'P';
const byte GET_PERISTENT = 'p';

const byte SET_LED_LEN = 5;
const byte SET_LCD_LEN = 22;

byte MsgBuffer[25];
bool DoUpdateLcd=false;

void SendMessage(byte cmd,byte chan,short val,bool x)
{
    Serial.write(cmd);
    Serial.write(chan);
    Serial.write(val%255);
    Serial.write(val/255);
}

void SendMessage(byte cmd,byte chan,byte val)
{
    Serial.write(cmd);
    Serial.write(chan);
    Serial.write(val);
}

void SendMessage(byte cmd,short chan)
{
    if(chan!=-1)
    {
        Serial.write(cmd);
        Serial.write(chan);
    }
}

void ReadSerial()
{
    if(Serial.available()>0)
    {
        MsgBuffer[0]=Serial.read();
        switch(MsgBuffer[0])
        {
            case SET_LED: {ParseL(); break;}
            case SET_LCD: {Parsel(); break;}
            case SYSTEM: {ParseS(); break;}
            case SET_PERISTENT: {ParseP(); break;}
            case GET_PERISTENT: {Parsep(); break;}
            default : { break;}
        }
    }
}

byte blockingRead()
{
    while (Serial.available()<1){}
    return Serial.read();
    
}

void blockingRead(byte* A,byte start,byte cnt)
{
    for (byte i = start; i < cnt+start; i++)
    {
        A[i]=blockingRead();
    }
    
}

void ParseL()
{
    blockingRead(MsgBuffer,1,4);
    for (byte i = 0; i < 3; i++)
    {
        LedValue[LedAlias[i][MsgBuffer[1]]][i]=MsgBuffer[i+2];
    }
}

void Parsel()
{
    blockingRead(MsgBuffer,1,21);
    for (byte i = 0; i < 20; i++)
    {
        LcdBuff[MsgBuffer[1]][i]=MsgBuffer[i+2];
    }
    LcdBuff[MsgBuffer[1]][20]='\0';
    DoUpdateLcd=true;
}

void ParseS()
{
    blockingRead(MsgBuffer,1,1);
    switch(MsgBuffer[1])
    {
        case SYSTEM_DISCOVER: {SendMessage(SYSTEM,RESPONCE_DISOVER,DEVICE_ID); Init=true; break;}
        case SYSTEM_RESET: {Lcd.clear();Lcd.print("RESET");delay(1000);software_Reset(); break;}
        default : { break;}
    }
}

void ParseP()
{
    blockingRead(MsgBuffer,1,2);
    WritePers(MsgBuffer[1],MsgBuffer[2]);
    RefreshPersistence();
}

void Parsep()
{
    blockingRead(MsgBuffer,1,1);
    SendMessage(RESPONCE_PERSISTENT,MsgBuffer[1],readPers(MsgBuffer[1]));
}

File: PHI Control Center/PHI Control Center/Classes/GHP.cs
using PHI_Control_Center.RTP;
using PluginBase;
using System;
using System.Collections.Generic;
using System.IO.Ports;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace PHI_Control_Center.Classes
{
    public class GHP : IGHP_Implementation
    {
        public const byte ANALOG_CHANGE = (byte)'A';
        public const byte BUTTON_PRESS = (byte)'B';
        public const byte BUTTON_RELEASE = (byte)'R';
        public const byte ENCODER_INCREMENT = (byte)'I';
        public const byte ENCODER_DECREMENT = (byte)'D';
        public const byte RESPONCE_PERSISTENT = (byte)'r';
        public const byte RESPONCE_DISOVER = (byte)'w';
        public const byte SYSTEM = (byte)'S';
        public const byte SYSTEM_DISCOVER = (byte)'W';
        public const byte SYSTEM_RESET = (byte)'R';
        public const byte SET_LED = (byte)'L';
        public const byte SET_LCD = (byte)'l';
        public const byte SET_PERISTENT = (byte)'P';
        public const byte GET_PERISTENT = (byte)'p';

        public SerialPort port;
        public ConnectedDevice device;
        public Dictionary<string, IPlugin> plugins;
        public byte[] MsgBuffer;

        public static byte DeviceConnectedToPort(string com)
        {
            SerialPort tmp = new SerialPort();
            tmp.BaudRate = 115200;
            tmp.PortName = com;
            tmp.ReadTimeout = 500;
            tmp.Open();
            tmp.DiscardInBuffer();
            tmp.Write(new byte[] { SYSTEM, SYSTEM_DISCOVER }, 0, 2);
            try
            {
                int Bc = 0;
                while (tmp.ReadByte() != RESPONCE_DISOVER && Bc<75) { Bc++; }
                if (Bc == 75)
                {
                    tmp.Close();
                    return 0;
                }
                byte res = (byte)tmp.ReadByte();
                tmp.Close();
                return res;
            }
            catch
            {
                tmp.Close();
                return 0;
            }
        }

        public static string DevicePort(byte ID)
        {
            foreach (string com in SerialPort.GetPortNames())
            {
                if (DeviceConnectedToPort(com) == ID) { return com; }
            }
            return "";
        }

        public static bool AddonEnabled(byte ID, byte AddonAddres)
        {
            string port = DevicePort(ID);
            if (port == "") { return false; }
            return GetPeristent(port, AddonAddres) == 1;
        }

        public static byte GetPeristent(string port, byte addres)
        {
            SerialPort tmp = new SerialPort();
            tmp.BaudRate = 115200;
            tmp.PortName = port;
            tmp.Open();
            tmp.DiscardInBuffer();
            tmp.Write(new byte[] { GET_PERISTENT, addres }, 0, 2);
            try
            {
                int Bc = 0;
                while (tmp.ReadByte() != RESPONCE_PERSISTENT && Bc < 75) { Bc++; }
                if (Bc == 75)
                {
                    tmp.Close();
                    return 0;
                }
                tmp.ReadByte();
                byte res = (byte)tmp.ReadByte();
                tmp.Close();
                return res;
            }
            catch
            {
                tmp.Close();
                return 0;
            }
        }

        public static void SetPeristent(string port, byte addres,byte value)
        {
            SerialPort tmp = new SerialPort();
            tmp.BaudRate = 115200;
            tmp.PortName = port;
            tmp.Open();
            tmp.DiscardInBuffer();
            tmp.Write(new byte[] { SET_PERISTENT, addres,value }, 0, 3);
            tmp.Close();
        }

        public GHP(string com,ConnectedDevice d,Dictionary<string,IPlugin> pl)
        {
            device = d;
            plugins = pl;
            port = new SerialPort();
            port.BaudRate = 115200;
            port.PortName = com;
            port.Open();
            MsgBuffer = new byte[25];
        }

        public void SendMessage(params byte[] bytes)
        {
            port.Write(bytes, 0, bytes.Length);
        }

        public void read()
        {
            if (port.BytesToRead > 0)
            {
                MsgBuffer[0] = blockingRead();
                switch (MsgBuffer[0])
                {
                    case ANALOG_CHANGE: { ParseA(); break; }
                    case BUTTON_PRESS: { ParseB(); break; }
                    case BUTTON_RELEASE: { ParseR(); break; }
                    case ENCODER_INCREMENT: { ParseI(); break; }
                    case ENCODER_DECREMENT: { ParseD(); break; }
                    case SYSTEM: { ParseS(); break; }
                    default: { break; }
                }
            }
        }

        public void ParseS()
        {
            blockingRead(MsgBuffer, 1, 1);
            switch (MsgBuffer[1])
            {
                case SYSTEM_RESET: { RTPthread.Stop(true); break; }
                default: { break; }
            }
        }

        public void ParseD()
        {
            blockingRead(MsgBuffer, 1, 1);
            ExecuteMapping(GetMapping(CType.Encoder, MsgBuffer[1]), MsgBuffer[1], 0);
        }

        public void ParseI()
        {
            blockingRead(MsgBuffer, 1, 1);
            ExecuteMapping(GetMapping(CType.Encoder, MsgBuffer[1]), MsgBuffer[1], 1);
        }

        private void ParseR()
        {
            blockingRead(MsgBuffer, 1, 1);
            ExecuteMapping(GetMapping(CType.Button, MsgBuffer[1]), MsgBuffer[1], 0);
        }

        public void ParseB()
        {
            blockingRead(MsgBuffer, 1, 1);
            ExecuteMapping(GetMapping(CType.Button, MsgBuffer[1]), MsgBuffer[1], 1);
        }

        public void ParseA()
        {
            blockingRead(MsgBuffer, 1, 3);
            int val = MsgBuffer[2] + (MsgBuffer[3]<<8);
            ExecuteMapping(GetMapping(CType.Analog, MsgBuffer[1]),MsgBuffer[1],val);
        }

        public byte blockingRead()
        {
            while (port.BytesToRead < 1) { }
            return (byte)port.ReadByte();

        }

        public void blockingRead(byte[] A, byte start, byte cnt)
        {
            for (byte i = start; i < cnt + start; i++)
            {
                A[i] = blockingRead();
            }

        }

        public Mapping GetMapping(CType type,byte id)
        {
            string bankname = device.Controlls[type][id].BankName;
            bankname = device.ActivePage[bankname];
            if (device.Controlls[type][id].Mappings.ContainsKey(bankname))
            {
                return device.Controlls[type][id].Mappings[bankname];
            }
            else
            {
                return null;
            }
        }

        private void ExecuteMapping(Mapping mapping,byte chan, int val)
        {
            if (mapping != null)
            {
                try
                {
                    plugins[mapping.PluginName].GetFunctions()[mapping.FunctionName].function(chan, val, mapping.args);
                }
                catch(Exception e)
                {
                    Console.WriteLine(e);
                }
            }
        }

        public void Close()
        {
            port.Close();
        }
    }
}
